### PR TITLE
Dual-Stack AO2 Client to handle both TCP and Websocket connections seemlessly

### PR DESCRIPTION
**Pull request by Salanto**
Is this a breaking change? Depends on how you view it.

If you intentionally sabotaged WebAO usage on your server by either limiting what Clients can join trough the websocket or just disabled them, this PR would break any connectivity we would have to that server till they either update or reenable websockets.

The replacement is a nice-to-have, as it finally removes the need for two ports and different connection medias between the regular AO2-Client and WebAO, giving new servers less of a headache implementing both and allowing already existing servers to streamline their connection code.

Closes https://github.com/AttorneyOnline/AO2-Client/issues/739

### CODE REVIEW

**Comment by stonedDiscord** [comment]
you can still sabotage webAO by using the version check packet

**Comment by Salanto** [comment]
> 

Yeah, but at that point is it even worth it anymore? Just allow WebAO, its not harmful.

**Comment by Crystalwarrior** [comment]
Do not remove tcp support or no one will update because their favorite server breaks or they can't play with friends who are not technically inclined.

Instead make it connect by websockets first, and if it fails fall back to tcp. Add settings that allow you to disable tcp entirely, and make it off by default.

**Comment by Salanto** [comment]
> Do not remove TCP support or no one will update because their favourite server breaks or they can't play with friends who are not technically inclined.
> 
> Instead make it connect by WebSockets first, and if it fails fall back to TCP. Add settings that allow you to disable TCP entirely, and make it off by default.

Literally what? Name me one of the popular servers that is not CaseCafe or DRO, which has no Websocket support in general and their mostly unique client, that don't have WebAO support. This is no harder to set up on the server owner as all major derivatives and servers have WebSocket support which is turned on with a single config change. 

Adding Websockets over TCP only adds more clutter to the AOClient with no tangible benefit besides "we backwards compatible because forcing owners to change a false to true is too much."

**Review by oldmud0 — CHANGES_REQUESTED** [review]
staying compatible with legacy tcp is like 5 lines. i will prove it to you

there are a number of servers that don't have ws forwarded properly or don't want to allow webao for whatever reason. They would not be prepared for this otherwise major change

**Comment by Salanto** [comment]
In the end I can't stop you from doing so.

**Comment by oldmud0** [comment]
oh crap, houston we got a big big problem. need to rebuild the windows qt5 builder so that it includes the websockets module

**Comment by Salanto** [comment]
Given that Github Actions produce usable clients, we don't need to run the same CI twice, but I am sure to be corrected because some people still use 32-bit operating systems in C++ and I know at least one person will complain.

**Comment by AwesomeAim** [comment]
This completely breaks favourites. Also, I can apparently double click on a favourite entry and it stores my double click and then the next time I click on a server that replies it immediately joins it.

File: include/networkmanager.h [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 #include <QDnsLookup>
 #include <QNetworkAccessManager>
-#include <QTcpSocket>
+#include <QtWebSockets/QWebSocket>
```
Comment (stonedDiscord) [inline include/networkmanager.h:9]: where did TCP go

**Review by stonedDiscord — CHANGES_REQUESTED** [review]

**Comment by Crystalwarrior** [comment]
> This completely breaks favourites. Also, I can apparently double click on a favourite entry and it stores my double click and then the next time I click on a server that replies it immediately joins it.

Just launched it, favorites seem to work?

**Comment by Crystalwarrior** [comment]
OH I have a clue. It only uses TCP backend when connecting to Favorites servers

File: src/networkmanager.cpp [hunk 2]
```diff
@@ -128,26 +131,98 @@ void NetworkManager::request_document(MSDocumentType document_type,
...
-  server_socket->abort();
+  disconnect_from_server();
 
   qInfo().nospace().noquote() << "connecting to " << p_server.ip << ":"
                               << p_server.port;
 
-  server_socket->connectToHost(p_server.ip, p_server.port);
+  switch (p_server.socket_type) {
+  default:
+      p_server.socket_type = TCP;
```
Comment (oldmud0) [inline src/networkmanager.cpp:141]:
```suggestion
    p_server.socket_type = TCP;
    [[fallthrough]];
```
clang-tidy had one job

File: src/text_file_functions.cpp [hunk 3]
```diff
@@ -274,9 +274,27 @@ QVector<server_type> AOApplication::read_serverlist_txt()
...
-        f_server.name = line_contents.at(2);
+        // This one gives us the socket type
+        if (line_contents.size() >= 4) {
+          QString socket_type = line_contents.at(0).toLower();
+          if (socket_type == "ws") {
+            f_server.socket_type = WEBSOCKETS;
+          }
+          else if (socket_type == "tcp") {
+            f_server.socket_type = TCP;
+          }
```
Comment (oldmud0) [inline src/text_file_functions.cpp:288]: consider changing this out for table heckin' lookup and/or moving the conversion code to datatypes.h next to the enum

File: src/aoapplication.cpp [hunk 4]
```diff
@@ -146,8 +146,17 @@ void AOApplication::add_favorite_server(int p_server)
...
 
   QString str_port = QString::number(fav_server.port);
+  QString network_type;
+  switch (fav_server.socket_type) {
+  default:
+    fav_server.socket_type = TCP;
+  case TCP:
+    network_type = "TCP";
+  case WEBSOCKETS:
+    network_type = "WS";
```
Comment (oldmud0) [inline src/aoapplication.cpp:158]: why does the write code use uppercase but the read code uses lowercase

Comment (Salanto) [reply src/aoapplication.cpp:158]: Sorry didn't read the read code will fix it.

**Comment by Salanto** [comment]
Should be good to go now. Tested on MingW 32-bit and the CI action.

File: include/networkmanager.h (hunk 1 above)
Comment (oldmud0) [reply include/networkmanager.h:9]: bye

**Review by oldmud0 — DISMISSED** [review]
this has waited far too long

**Review by oldmud0 — APPROVED** [review]
this is a major feature for 2.10, so let's get this merged

File: include/networkmanager.h (hunk 1 above)
Comment (oldmud0) [reply include/networkmanager.h:9]: it's probably being included transitively by another include, I wouldn't worry about it too much.

File: src/text_file_functions.cpp [hunk 5]
```diff
@@ -278,6 +278,16 @@ QVector<server_type> AOApplication::read_serverlist_txt()
...
+
+        if (line_contents.size() >= 4) {
+          QString socket_type = line_contents.at(3).toLower();
+          if (socket_type == "ws" || socket_type == "tcp") {
+            f_server.socket_type = to_connection_type.value(socket_type);
+          }
+          else {
+            qWarning() << "Serverlist.txt - Unknown socket type: " << socket_type;
+          }
+        }
```
Comment (oldmud0) [inline src/text_file_functions.cpp:290]: Fails to support colons in server names.

**Review by oldmud0 — CHANGES_REQUESTED** [review]

File: src/text_file_functions.cpp (hunk 5 above)
Comment (Salanto) [reply src/text_file_functions.cpp]: Latest changes now support it : 
image

**Review by stonedDiscord — APPROVED** [review]

**Pull request merged**